main.go

```diff
@@ -18,40 +25,74 @@ import (
 )
 
 func main() {
+	log.SetFlags(log.Flags() | log.Lshortfile)
+	log.SetOutput(os.Stderr)
+
+	if err := run(); err != nil {
+		log.Fatalln("main failed to run:", err)
+	}
+
+	log.Println("sucessful shutdown")
+}
+
+func run() error {
 	config, err := LoadConfig()
 	if err != nil {
-		panic(err)
+		return fmt.Errorf("failed to LoadConfig: %w", err)
 	}
 
-	gin.SetMode(config.Env)
-
 	// migrate the db on startup
 	m, err := migrate.New("file://sql", config.DatabaseURL)
 	if err != nil {
-		panic(err)
+		return fmt.Errorf("failed to migrate.New: %w", err)
 	}
 
-	m.Up()
+	if err := m.Up(); err != nil {
+		if !errors.Is(err, migrate.ErrNoChange) {
+			return fmt.Errorf("failed to migrate Up: %w", err)
+		}
+
+		log.Println("no new migrations detected, schema is current")
+	}
 
 	db, err := sqlx.Open("postgres", config.DatabaseURL)
 	if err != nil {
-		panic(err)
+		return fmt.Errorf("failed to sqlx.Open: %w", err)
 	}
 
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	wg := sync.WaitGroup{}
+
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
+		ticker := time.NewTicker(time.Hour)
+		defer ticker.Stop()
```

Comment:
Since we're using `cancel` to shutdown this goroutine, is there benefit in deferring this and the `wg.Done()` vs just calling them explicitly before the `return` below in the `ctx.Done()` select?

Seems like the behavior would be about the same, the main difference being locality of behavior. Not deferring them means all the behavior associated with "this goroutine is done" is all group under that `ctx.Done()` select.

Thoughts? Is there something about `defer` that adds extra value here?

Reply:
This is a good question, and I have two reasons for why I chose to defer these instead of placing them in the `ctx.Done()` case.

1. I like to defer cleanup as quickly as possible after their creation. This has the benefit of ensuring they are called in case exit conditions are added or changed in the future, we don't need to remember to do all of the shutdown again. We also don't need to remember the order the cleanup steps need to be run in, which defer handles for us.
2. We definitely want these to run, no matter what, before the goroutine returns, even _if there is a panic_.

Here's an example of this in action:
```go
package main

import "log"

func main() {
	defer func() {
		log.Println("i'm logging before the panic kills me")
	}()

	panic("at the disco")
}
```

Or checkout this blog: https://go.dev/blog/defer-panic-and-recover
> When the function F calls panic, execution of F stops, any deferred functions in F are executed normally, and then F returns to its caller. 

Are we in danger of a panic here? Eh, probably not, but thems are my thoughts. 👍

Reply:
> I like to defer cleanup as quickly as possible after their creation.

Good point! I guess that's a different aspect of locality in this situation.

Also good call on `defer`'d functions being called even in case of a panic, that slipped my mind and is definitely valuable! You've really got a good grasp on go language semantics 😄 💯